frontend/middle-3/fp.md

```diff
@@ -0,0 +1,183 @@
+* What is the difference between Sum Types and Unions?
+* Why are Sums and Products called so?
+* Why AlgDT are called also inductive data types?
+* Why Sum Types are called also tagged union? What is a tag there?
+* How algebraic properties respected in AlgDT (like Symmetry, Identity, Associativity and Distributivity)?
+* What is a difference in tracking nullable values with Unions and with Sum Types?
+* Why does AlgDT fit well for the abstract syntax tree representation?
+* Why AlgDT with pattern-matching is more type safe approach than Unions?
+* Why is exhaustiveness checking possible for AlgDT but impossible for objects with class hierarchy?
+* What are drawbacks of the AlgDT comparing with Unions?
+* What are smart constructors for AlgDT? What are destructors?
+* Which invariants can be modeled with AlgDT and which cannot?
+
+### Resources:
+* [Better JS Cases with Sum Types](https://medium.com/fullstack-academy/better-js-cases-with-sum-types-92876e48fd9f)
+* [An Introduction to ADTs and Structural Pattern Matching in TypeScript](https://blog.carbonfive.com/2018/01/08/an-introduction-to-adts-and-structural-pattern-matching-in-typescript/)
+* [TypeScript: Precise domain modeling with discriminated unions](https://codewithstyle.info/typescript-discriminated-union-types/)
+* [Algebraic Data Types explained in Statically typed Javascript](https://medium.com/@notgiorgi/algebraic-data-types-explained-in-statically-typed-javascript-4ad31c2b12c9)
+* [Algebraic data type on Wikipedia](https://en.wikipedia.org/wiki/Algebraic_data_type)
+* [Null-tracking, or the difference between union and sum](https://waleedkhan.name/blog/union-vs-sum-types/)
+* [An Introduction to Algebraic Data Types](https://itnext.io/an-introduction-to-algebraic-data-types-9429e49eac27)
+* [Talk] [Domain Modeling Made Functional - Scott Wlaschin](https://www.youtube.com/watch?v=Up7LcbGZFuo)
+* [Easy domain modelling with types](http://blog.ploeh.dk/2016/11/28/easy-domain-modelling-with-types/)
+* [Алгебраические типы данных и их использование в программировании](http://fprog.ru/2009/issue2/roman-dushkin-algebraic-data-types/)
+* [(*) The algebra (and calculus!) of algebraic data types types](https://codewords.recurse.com/issues/three/algebra-and-calculus-of-algebraic-data-types)
+
+## Combinator library
+* What is a combinator library?
+* What are primitives and combinators?
+* What is a closure property and why is it useful in combinator libraries?
+* Why is parsing a greatly appropriate domain for modeling with combinators?
+* How to determine that domain is a good for combinators?
+  * If it is too narrow it will not a flexible, so it is simpler to consider just a straightforward contract.
+  * If too wide, you can meet the case where you cannot compose the combinator results. so it is still possible to combine data and not get conflicts
+* Is it possible to save performance for combinators application results comparing to using simple primitives?
```

Comment:
Is it possible to achieve better performance for application using combinators instead of simple primitives?